### DISCUSSION TITLE
How to simply remove a variation, or add a variation

### DISCUSSION THREAD

**Discussion by tophattwaffle** [etsy/open-api#833] [Q&A] [answered]
Hello,

I have my application connecting to Etsy V3 API and can hit end points as needed. After reading through the documentation - I wanted to see if there was some simpler way to handle what I want to do.

Using one of my listings as an example, it has Primary Color and Secondary Color that share the same list of 15 colors. There are a few listings I have that share these same colors, so updating the variation colors all at once across all listings as I get stock in and out would be wonderful. 

My understanding on how to remove a color would be as follows:

1. Hit the getListingInventory endpoint.
2. Find the variations I want to remove and remove them from the returned data.
3. Send the new data back Etsy using the updateListingInventory endpoint.

Is the above 3 steps the general idea to update/add/remove variations?

I first started by getting a listing inventory, and then sending the JSON right back to the `updateListingInventory` endpoint. I figured if the thought process I outlined above was right (and what I was reading on other discussions) this would work. When I do this I get `{"error":"Array contains invalid keys: product_id,is_deleted"}`. Would I be correct to assume I need to remove these properties from the json data before sending it back to update? Are there any other properties that need to be removed?

Am I correct in thinking that to remove a variation, I remove associated object in the "products" array in the json structure, then once again send it back to the update endpoint?

Finally, when adding a variation, I noticed that there is a product for every possible combination of variations. My variations have no costs or or quantity associated with them. I am assuming that is why there is every possible combination listed.

When I want to add a variation to a listing, what properties do I actually need to add to the listingInventory before I send it back to to update endpoint?

Do I also need to generate every possible combination before sending it back to the update endpoint?

The documentation/tutorial page isn't very clear on my questions, but otherwise seems to be decent.

**Comment by hyyyper** [comment, marked as answer]
Hi tophattwaffle,

Your order of operations is correct 👍.

As you discovered, there is a discrepancy between the data that is returned by getListingInventory and the data you are allowed to send to updateListingInventory.

The JSON that you send to the update endpoint can ONLY contain the data listed in the docs:
https://developers.etsy.com/documentation/reference/#operation/updateListingInventory

I find it easiest to loop over the GET data and reconstruct it variant by variant, including any changes you want to make.

Something like this:
```js
// loop over the data
// append each variant to products
const this_variant = {
    sku: "YOUR_SKU",
    offerings: [{
        price: 33.33,
        quantity: 123,
        is_enabled: true,
    }],
    property_values: [
        ...
    ]
}
```

You are correct about removing a variation. You just send the products array back without that variant.
You also do need to send every possible combination of options everytime, whether or not they are active variants.
